### PR TITLE
(968c) Show users name instead of username on programme history entries

### PR DESCRIPTION
**Pull request by jsrobertson**
## Context

We want to display the name of the user who has added a programme history record.

## Changes in this PR
- Updated the User factory to return randomly formatted name to replicate what the manage-users-api might return.
- Add a new `presentCourseParticipation` method to `CourseService` which can be called from controllers where we need to return a formatted name for the `addedBy` user.

### Before
image

### After
image

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/controllers/refer/referralsController.test.ts [hunk 1]
```diff
@@ -200,13 +209,33 @@ describe('ReferralsController', () => {
...
         latestCourseParticipation,
         earliestCourseParticipation,
       ])
+      const presentedLatestCourseParticipation: CourseParticipationPresenter = {
+        ...latestCourseParticipation,
+        addedByDisplayName: StringUtils.convertToTitleCase(addedByUser.name),
+      }
+      courseService.presentCourseParticipation
+        .mockResolvedValueOnce(presentedEarliestCourseParticipation)
+        .mockResolvedValueOnce(presentedLatestCourseParticipation)
```
Comment (yndajas) [inline server/controllers/refer/referralsController.test.ts:236]: This relies on `presentCourseParticipation` being called only twice and in this specific order, right?

Probably fine given it's specific to this test, but using `jest-when` could avoid any confusion if we rearrange anything here (e.g. if we switch to reverse sorting). Something like (with the arguments flipped and the presenter naming convention update) `when(courseService.presentCourseParticipation).calledWith(token, earliestCourseParticipation).mockResolvedValue(earliestCourseParticipationPresenter)`

Same in the other controller

File: server/controllers/refer/courseParticipationsController.test.ts [hunk 2]
```diff
@@ -247,10 +265,22 @@ describe('CourseParticipationsController', () => {
...
-    const earliestCourseParticipation = courseParticipationFactory.build({ createdAt: '2022-01-01T12:00:00.000Z' })
+    const earliestCourseParticipation = courseParticipationFactory.build({
+      addedBy: addedByUser.username,
+      createdAt: '2022-01-01T12:00:00.000Z',
+    })
+    const presentedEarliestCourseParticipation: CourseParticipationPresenter = {
+      ...earliestCourseParticipation,
+      addedByDisplayName: StringUtils.convertToTitleCase(addedByUser.name),
+    }
     const latestCourseParticipation = courseParticipationFactory.build({ createdAt: '2023-01-01T12:00:00.000Z' })
```
Comment (yndajas) [inline server/controllers/refer/courseParticipationsController.test.ts:279]: Does the `addedBy` need adding to the `latestCourseParticipation` too?

**Review by yndajas — COMMENTED** [review]
Looks good! Just a few suggestions for tidy ups

File: server/utils/courseParticipationUtils.ts [hunk 3]
```diff
@@ -109,7 +110,7 @@ export default class CourseParticipationUtils {
   }
 
   static summaryListOptions(
-    courseParticipation: CourseParticipation,
+    presentedCourseParticipation: CourseParticipationPresenter,
```
[the remark this answers is not in the document]
Comment (yndajas) [reply server/utils/courseParticipationUtils.ts:113]: I can see this is all over the changes, so I wouldn't worry too much about it if it's a pain to change everywhere!

File: server/services/courseService.ts [hunk 4]
```diff
@@ -1,17 +1,23 @@
 import createError from 'http-errors'
 import type { ResponseError } from 'superagent'
 
-import type { CourseClient, RestClientBuilder } from '../data'
+import type UserService from './userService'
+import { type CourseClient, type RestClientBuilder } from '../data'
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/services/courseService.ts:5]: Not quite sure actually, good spot though.  Will revert.

File: server/services/index.ts [hunk 5]
```diff
@@ -17,11 +17,11 @@ import {
 } from '../data'
 
 const services = () => {
-  const courseService = new CourseService(courseClientBuilder)
   const organisationService = new OrganisationService(prisonClientBuilder)
   const personService = new PersonService(hmppsAuthClientBuilder, prisonerClientBuilder)
   const referralService = new ReferralService(referralClientBuilder)
   const userService = new UserService(hmppsManageUsersClientBuilder, caseloadClientBuilder)
+  const courseService = new CourseService(courseClientBuilder, userService)
```
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/services/index.ts:24]: Unfortunately not due to `Variable 'userService' is used before being assigned.` However I did wonder and discuss with @Gweaton if we could see this becoming a problem in future if services start to require other services more regularly.  The alternative was to create a `new UserService` inside `CourseService` but decided to go down this route for now.

File: server/utils/courseParticipationUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsrobertson) [reply server/utils/courseParticipationUtils.ts:113]: I was torn between  `presentedCourseParticipation` and `courseParticipationPresenter` so happy to change and go with `courseParticipationPresenter`.

File: server/services/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (yndajas) [reply server/services/index.ts:24]: Ah of course. Yeah that's true, we could end up with cyclical dependencies. I guess we can be cautious about adding more inter-service dependencies

**Review by yndajas — APPROVED** [review]

**Pull request merged**